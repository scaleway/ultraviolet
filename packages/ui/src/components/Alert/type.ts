export const ALERT_SENTIMENTS = [
  'danger',
  'info',
  'success',
  'warning',
  'neutral',
] as const

export type AlertSentiment = (typeof ALERT_SENTIMENTS)[number]
