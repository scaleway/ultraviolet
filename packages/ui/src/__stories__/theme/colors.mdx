import { Meta } from '@storybook/addon-docs/blocks'
import Colors from '../components/Colors'

<Meta title="Customization/Colors" />

# Customize colors

You can customize colors by extending the theme. <br />
This means default colors will apply if you didn't override them all.

To customize theme you will need to use `extendTheme` utility function. It will generate a well formatted theme to pass
to your Theme Provider.

You will need to pass an object with the same structure as the [default theme](https://github.com/scaleway/ultraviolet/blob/main/src/theme/tokens/light.ts), but with your own colors.

```tsx
import { ThemeProvider } from '@emotion/react'
import { Badge, extendTheme } from '@ultraviolet/ui'

function App() {
  const myTheme = extendTheme({
      colors: {
          danger: {
              background: '#ffe1e7',
              backgroundDisabled: '#fff5f7',
              backgroundHover: '#ffe1e7',
              backgroundStrong: '#dd3252',
              backgroundStrongDisabled: '#fbbac6',
              backgroundStrongHover: '#a6102d',
              border: '#a6102d',
              borderDisabled: '#fbbac6',
              borderHover: '#a6102d',
              borderStrong: '#ffffff',
              borderStrongDisabled: '#fbbac6',
              borderStrongHover: '#a6102d',
              text: '#a6102d',
              textDisabled: '#fbbac6',
              textHover: '#a6102d',
              textStrong: '#ffffff',
              textStrongDisabled: '#fff5f7',
              textStrongHover: '#ffffff',
          },
          ...,
      }
  })

  return (
    <ThemeProvider theme={myTheme}>
        <Badge sentiments="danger">Example</Badge>
    </ThemeProvider>
  )
}

export default App
```

### API

#### `extendTheme({extendedTheme}: {extendedTheme: Theme})`

This function take in parameters an object with de same structure of the default theme. It will return a deep merged object with the default theme.

## Default Values

Here are all default colors we use in Ultraviolet UI. You can switch to dark theme on storybook to see dark mode colors.

<Colors />
