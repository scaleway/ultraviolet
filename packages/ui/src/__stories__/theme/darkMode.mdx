import { Meta } from '@storybook/addon-docs/blocks'

<Meta title="Customization/Dark mode" />

# Dark mode

## How to?

Ultraviolet UI is designed to be used in both light and dark mode. The kit also provide both light and dark theme. <br />
You can enable easily switch between light and dark mode by passing `theme` or `darkTheme` to the `ThemeProvider` component.

Example:

```tsx
import { ThemeProvider } from '@emotion/react'
import { theme, darkTheme, Button, Text } from '@ultraviolet/ui'
import { useState } from 'react'

function App() {
  const [isDarkTheme, setIsDarkTheme] = useState(false)

  return (
    <ThemeProvider theme={isDarkTheme ? darkTheme : theme}>
      <Text as="p" variant="body">
        Current theme: {theme}
      </Text>

      <Button onClick={() => setIsDarkTheme(!isDarkTheme)}>Switch theme</Button>
    </ThemeProvider>
  )
}

export default App
```
