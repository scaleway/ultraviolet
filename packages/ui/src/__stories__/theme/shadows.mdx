import { Meta } from '@storybook/addon-docs/blocks'

<Meta title="Customization/Shadows" />

# Shadows

The default theme has a set of shadows that can be used to create a sense of depth and hierarchy.

## Customize shadows

You may also want to customize your shadows. To do that you need to create your own theme and pass it to the `ThemeProvider` component. <br />
You will need to pass an object with the same structure as the [default theme](https://github.com/scaleway/ultraviolet/blob/main/src/theme/tokens/light.ts), but with your own typography.

```tsx
import { ThemeProvider } from '@emotion/react'
import { Tooltip, extendTheme } from '@ultraviolet/ui'

function App() {
  const myTheme = extendTheme({
    shadows: {
      tooltip: '10px 0px 20px 16px #000000',
    },
  })

  return (
    <ThemeProvider theme={myTheme}>
      <Tooltip text="I'm a tooltip">
        <p>Example</p>
      </Tooltip>
    </ThemeProvider>
  )
}

export default App
```
