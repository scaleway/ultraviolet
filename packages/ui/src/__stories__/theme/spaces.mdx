import { Meta } from '@storybook/addon-docs/blocks'
import SpaceTable from './assets/space-table.png'
import SizingTable from './assets/sizing-table.png'
import lightTheme from '../../theme'
import { Card, Alert } from '@ultraviolet/ui'

<Meta title="Customization/Dimensions" />

# Dimensions

## Concept

In the theme tokens you will find <b>2 major dimension categories</b> for spacing `space` and `sizing`.
<br />
### Space

Our spacing is structured on the <b>8px</b> base unit. All the keys are based on this unit and are multiples of it. This allows for a consistent spacing scale that can be used across the application.
<br/>
Space is used for spacing <b>paddings</b>, <b>margins</b> and <b>gaps</b>. It is a scale of values that can be used to create consistent spacing between elements.
<br/>
<img src={SpaceTable} height="700" alt="" />
<br />
You can access to space in Ultraviolet using the theme:

```tsx
const MyComponent = styled.div`
  margin: ${({ theme }) => theme.space['1']}; // 0.5rem or 8px
`
```

<Card>
| **Do** | **Don't** |
|--------|-----------|
| âœ… **Do**: Use `space` for setting spacing (padding, margin, gap) to maintain consistency. | ðŸš« **Don't**: Use `space` on height, max-height, width, max-width. Use `sizing` for that. |
| | ðŸš« **Don't**: Use `space` on border radius, `theme.radii` exists for that |
| | ðŸš« **Don't**: Use `space` on font size and line height, `theme.typography` exists for that or use `<Text />` component |
</Card>


### Sizing

Sizing is a of 8px just like space. In order to differenciate space and sizing and avoid usage confusion we create different keys for sizing. Those keys are named by the multiplier * 1000. Multiplier being the number of 8px units.
<br />
Sizing is used mostly for <b>height</b> and <b>width</b> of elements. It is a scale of values that can be used to create consistent sizing of elements.
<br/>
<img src={SizingTable} height="1100" alt="" />
<br />

<Alert title="Be aware of deprecated tokens" sentiment="warning">Tokens that have yellow background in the above picture are deprecated and should be avoided as much as possible. We will remove them in a future major.</Alert>
<br />
You can access to sizing in Ultraviolet using the theme:

```tsx
const MyComponent = styled.div`
  margin: ${({ theme }) => theme.sizing['200']}; // 1rem or 16px
`
```

<Card>
| **Do** | **Don't** |
|--------|-----------|
| âœ… **Do**: Use `sizing` for setting width and height (min and max included) to maintain consistency. | ðŸš« **Don't**: Use `sizing` on padding, margin, gap, and any other spacing element. Use `space` for that. |
| | ðŸš« **Don't**: Use `sizing` on border and box shadow, we want to keep those in px and not in rem |
| | ðŸš« **Don't**: Use `sizing` on border radius, `theme.radii` exists for that |
| | ðŸš« **Don't**: Use `sizing` on font size and line height, `theme.typography` exists for that or use `<Text />` component |
</Card>
<br />

### Customize

You can customize the space and sizing by passing a custom theme to the `ThemeProvider`:

```tsx
import { ThemeProvider } from '@emotion/react'
import { Button, extendTheme } from '@ultraviolet/ui'

function App() {
  const myTheme = extendTheme({
    space: {
        0: '0rem',
        1: '0.5rem',
        2: '1rem',
        3: '1.5rem',
        4: '2rem',
        5: '2.5rem',
        6: '3rem',
        7: '3.5rem',
        8: '4rem',
        ...
    },
    sizing: {
        0: '0rem',
        100: '0.5rem',
        125: '0.625rem',
        150: '0.75rem',
        175: '0.875rem',
        200: '1rem',
        250: '1.25rem',
        ...
    }
  })

  return (
    <ThemeProvider theme={myTheme}>
        <Button>
            Example
        </Button>
    </ThemeProvider>
  )
}

export default App
```
