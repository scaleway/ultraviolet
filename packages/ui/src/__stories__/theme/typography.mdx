import { Meta } from '@storybook/addon-docs/blocks'

<Meta title="Customization/Typography" />

# Typography

In general way you won't have to use typography from theme as you may use Text component for any text. <br />
But if you need to use it, you can use it like this: `theme.typography`.

## Customize typography

You may also want to customize your typography fonts or sizes. To do that you need to create your own theme and pass it to the `ThemeProvider` component. <br />
You will need to pass an object with the same structure as the [default theme](https://github.com/scaleway/ultraviolet/blob/main/src/theme/tokens/light.ts), but with your own typography.

```tsx
import { css, Global, ThemeProvider } from '@emotion/react'
import { Text, extendTheme } from '@ultraviolet/ui'
import AsapRegularWoff2 from 'assets/fonts/asap/Asap-Regular.woff2'

const fonts = css`
  @font-face {
    font-family: 'Asap';
    font-style: normal;
    src: url(${AsapRegularWoff2}) format('woff2');
    font-weight: 400;
    font-display: swap;
  }
`

function App() {
  const myTheme = extendTheme({
    typography: {
      myTypography: {
        fontFamily: 'Asap',
        fontSize: '16px',
        fontWeight: 'Regular',
        letterSpacing: '0',
        lineHeight: '24px',
        paragraphSpacing: 'none',
        textCase: 'none',
        textDecoration: 'none',
        weight: '400',
      },
    },
  })

  return (
    <ThemeProvider theme={myTheme}>
      <Global styles={[fonts]} />
      <Text as="p" variant="myTypography">
        Example
      </Text>
    </ThemeProvider>
  )
}

export default App
```
