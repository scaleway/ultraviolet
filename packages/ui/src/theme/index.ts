import { consoleDarkTheme, consoleLightTheme } from '@ultraviolet/themes'
import deepmerge from 'deepmerge'
import ThemeRegistry from './ThemeRegistry'

export type ScreenSize = keyof typeof consoleLightTheme.screens

/**
 * @deprecated use UltravioletUITheme instead
 */
type SCWUITheme = typeof consoleLightTheme
type UltravioletUITheme = typeof consoleLightTheme

const { colors, shadows, typography, space, radii, screens } = consoleLightTheme

type RecursivePartial<T> = {
  [P in keyof T]?: RecursivePartial<T[P]>
}

/**
 * Will extend theme with new theme properties
 * @param {UltravioletUITheme} baseTheme the theme you want to extend from, by default it is set to light theme
 * @param {RecursivePartial<UltravioletUITheme>} extendedTheme the properties of a new theme you want to apply from baseTheme
 */
const extendTheme = (extendedTheme: RecursivePartial<UltravioletUITheme>) =>
  deepmerge(consoleLightTheme, extendedTheme) as UltravioletUITheme

// This type exclude overlay color
type Color = Extract<
  keyof typeof consoleLightTheme.colors,
  | 'primary'
  | 'secondary'
  | 'neutral'
  | 'success'
  | 'danger'
  | 'warning'
  | 'info'
>

type ExtendedColor = Color | 'white' | 'black'

const SENTIMENTS = [
  'primary',
  'secondary',
  'neutral',
  'success',
  'danger',
  'warning',
  'info',
] as const

const SENTIMENTS_WITHOUT_NEUTRAL = SENTIMENTS.filter(
  sentiment => sentiment !== 'neutral',
)

export type { SCWUITheme, UltravioletUITheme, Color, ExtendedColor }

export {
  colors,
  shadows,
  space,
  radii,
  screens,
  consoleDarkTheme as darkTheme,
  extendTheme,
  SENTIMENTS,
  SENTIMENTS_WITHOUT_NEUTRAL,
  typography,
  ThemeRegistry,
}

export default consoleLightTheme
