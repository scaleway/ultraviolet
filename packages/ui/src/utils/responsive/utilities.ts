import type { ScreenSize } from '../../theme'
import { screens } from '../../theme'

export const up = (size: ScreenSize, rules: string): string => {
  if (!(size in screens)) {
    return ''
  }

  return `@media (min-width: ${screens[size]}px) { ${rules} }`
}

export const down = (size: ScreenSize, rules: string): string => {
  if (!(size in screens)) {
    return ''
  }

  return `@media (max-width: calc(${screens[size]} - 1px)) { ${rules} }`
}
