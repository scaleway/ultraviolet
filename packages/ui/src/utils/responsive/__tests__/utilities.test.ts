/* eslint-disable eslint-comments/disable-enable-pair */

import { describe, expect, test } from 'vitest'
import type { ScreenSize } from '../../../theme'
import { screens } from '../../../theme'
import { down, up } from '../utilities'

describe('responsive utilities', () => {
  describe('up', () => {
    ;(Object.keys(screens) as ScreenSize[]).forEach(brk => {
      test(`works with ${brk}`, () => {
        expect(up(brk, 'border: 1px solid red;')).toMatchSnapshot()
      })
    })
    test('should return null when unknown size', () => {
      // @ts-expect-error we send the wrong screen size on purpose
      expect(up('yolo', 'border: 1px solid red;')).toMatchSnapshot()
    })
  })
  describe('down', () => {
    ;(Object.keys(screens) as ScreenSize[]).forEach(brk => {
      test(`works with ${brk}`, () => {
        expect(down(brk, 'border: 1px solid red;')).toMatchSnapshot()
      })
    })
    test('should return null when unknown size', () => {
      // @ts-expect-error we send the wrong screen size on purpose
      expect(down('yolo', 'border: 1px solid red;')).toMatchSnapshot()
    })
  })
})
