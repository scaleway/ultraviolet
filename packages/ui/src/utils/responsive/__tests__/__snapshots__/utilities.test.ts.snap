// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`responsive utilities > down > should return null when unknown size 1`] = `""`;

exports[`responsive utilities > down > works with large 1`] = `"@media (max-width: calc(1440px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > down > works with medium 1`] = `"@media (max-width: calc(1280px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > down > works with small 1`] = `"@media (max-width: calc(1080px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > down > works with xlarge 1`] = `"@media (max-width: calc(1800px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > down > works with xsmall 1`] = `"@media (max-width: calc(720px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > down > works with xxsmall 1`] = `"@media (max-width: calc(320px - 1px)) { border: 1px solid red; }"`;

exports[`responsive utilities > up > should return null when unknown size 1`] = `""`;

exports[`responsive utilities > up > works with large 1`] = `"@media (min-width: 1440pxpx) { border: 1px solid red; }"`;

exports[`responsive utilities > up > works with medium 1`] = `"@media (min-width: 1280pxpx) { border: 1px solid red; }"`;

exports[`responsive utilities > up > works with small 1`] = `"@media (min-width: 1080pxpx) { border: 1px solid red; }"`;

exports[`responsive utilities > up > works with xlarge 1`] = `"@media (min-width: 1800pxpx) { border: 1px solid red; }"`;

exports[`responsive utilities > up > works with xsmall 1`] = `"@media (min-width: 720pxpx) { border: 1px solid red; }"`;

exports[`responsive utilities > up > works with xxsmall 1`] = `"@media (min-width: 320pxpx) { border: 1px solid red; }"`;
